src/lock/update.rs

```diff
@@ -0,0 +1,233 @@
+use crate::dependencies::{FrozenDependency, LockedDependency, SimpleDependency};
+use crate::operations::UpdateOptions;
+
+#[derive(Debug)]
+pub struct Updater {
+    python_version: Option<String>,
+    sys_platform: Option<String>,
+}
+
+impl Updater {
+    pub fn new() -> Self {
+        Updater {
+            python_version: None,
+            sys_platform: None,
+        }
+    }
+
+    pub fn set_options(&mut self, update_options: UpdateOptions) {
+        let UpdateOptions {
+            python_version,
+            sys_platform,
+        } = update_options;
+        self.python_version = python_version;
+        self.sys_platform = sys_platform;
+    }
+
+    /// Applies a set of new FrozenDependency to the lock
+    // Basically, update `self.dependencies` using the new frozen deps,
+    // making sure no existing information in the lock is lost.
+    // Note that we only modify existing lines or add new ones
+    // (no deletion occurs).
+    pub fn update(
+        &self,
+        locked_dependencies: &mut Vec<LockedDependency>,
+        frozen_dependencies: &[FrozenDependency],
+    ) {
+        self.patch_existing_deps(locked_dependencies, frozen_dependencies);
+        self.add_missing_deps(locked_dependencies, frozen_dependencies);
+    }
+
+    /// Add dependencies from `frozen_deps` that were missing in the lock
+    fn add_missing_deps(
+        &self,
+        locked_dependencies: &mut Vec<LockedDependency>,
+        frozen_deps: &[FrozenDependency],
+    ) {
+        let known_names: Vec<_> = locked_dependencies.iter().map(|d| d.name()).collect();
+        let new_deps: Vec<_> = frozen_deps
+            .iter()
+            .filter(|x| !known_names.contains(&&x.name))
```

Comment:
Why the double ref? (I mean, it's *technically* the exact type, but I'm not sure it's necessary!)

Reply:
Don't know. I guess it used to not compile some time ago